src/ui/public/indices/get_indices.js

```diff
@@ -1,34 +1,77 @@
 import { pluck, reduce, size } from 'lodash';
 
+const getIndexNamesFromAliasesResponse = response => {
+  if (response.status === 404) {
+    return [];
+  }
+
+  return reduce(response, (list, { aliases }, indexName) => {
+    if (size(aliases) > 0) {
+      list.push(...Object.keys(aliases).map(alias => ({
+        name: alias,
+      })));
+    }
+
+    return list;
+  }, []);
+};
+
+const getIndexNamesFromSearchResponse = response => {
+  if (
+    response.status === 404
+    || !response.aggregations
+  ) {
+    return [];
+  }
+
+  return response.aggregations.indices.buckets.map(bucket => ({
+    name: bucket.key,
+    docCount: bucket.doc_count,
+  }));
+};
+
 export function IndicesGetIndicesProvider(esAdmin) {
-  const getIndexNamesFromAliasesResponse = json => {
-    // Assume this function won't be called in the event of a 404.
-    return reduce(json, (list, { aliases }, indexName) => {
-      list.push(indexName);
-      if (size(aliases) > 0) {
-        list.push(...Object.keys(aliases));
-      }
-      return list;
-    }, []);
-  };
+  return async function getIndices(indexPattern, maxNumberOfMatchingIndices) {
+    if (!indexPattern) {
+      throw new Error('Please provide an indexPattern string to getIndices().');
+    }
+
+    if (!maxNumberOfMatchingIndices || maxNumberOfMatchingIndices < 0) {
+      throw new Error('Please provide a maxNumberOfMatchingIndices value greater than 0 to getIndices().');
+    }
 
-  const getIndexNamesFromIndicesResponse = json => {
-    if (json.status === 404) {
+    // Cross-cluster search query missing the index pattern part.
+    if (!indexPattern.split(':')[1]) {
       return [];
     }
 
-    return pluck(json, 'index');
-  };
+    // const aliasesResponse = await esAdmin.indices.getAlias({
+    //   index: indexPattern,
+    //   allowNoIndices: true,
+    //   ignore: 404,
+    // });
 
-  return async function getIndices(query) {
-    const aliases = await esAdmin.indices.getAlias({ index: query, allowNoIndices: true, ignore: 404 });
+    const aliases = [];//getIndexNamesFromAliasesResponse(aliasesResponse);
 
-    // If aliases return 200, they'll include matching indices, too.
-    if (aliases.status === 404) {
-      const indices = await esAdmin.cat.indices({ index: query, format: 'json', ignore: 404 });
-      return getIndexNamesFromIndicesResponse(indices);
-    }
+    const searchResponse = await esAdmin.search({
+      index: indexPattern,
+      ignore: [404],
+      allow_no_indices: true,
```

Comment:
Adding `terminate_after: 1` here will make sure that es only reads a single document from each index, meaning that the search won't need to do more work than necessary and should make this as performant as possible.